### ISSUE TITLE
Is it possible to read gzip while ignoring the CRC32?

### ISSUE THREAD

**Issue by golddranks** [alexcrichton/flate2-rs#117]
I'm doing batch processing with a lot of data, which will be validated later anyway. I'm using flate2's gzip functionality to uncompress this data. Profiling showed that the "mz_crc32" function is a significant time hog. Is it possible to read gzip while ignoring CRC32? I noticed from the source code that the GzDecoder wraps a CrcReader. Is there a version that doesn't calculate the CRC?

**Comment by alexcrichton** [comment]
Currently it's not possible to disable this, but the functionality could be added! I'd be surprised though if this dominted the actual decompression...

**Comment by golddranks** [comment]
I'll fork and do some experiments. Do you have any kind of API in mind, like exposing the option in a builder pattern or something?

**Comment by alexcrichton** [comment]
I'd imagine this probably just being a flag on the gz decoders. The flag would just disable feeding it internally to the crc checksum calculator. I'd be interested though in the results of more profiling (or some example files I could work with locally)

**Comment by golddranks** [comment]
I did some profiling, and witnessed a 16 seconds → 11 seconds reduction when not using the CRC reader. Here's some flamegraphs:

![2017-08-10 15 34 51](https://user-images.githubusercontent.com/2675542/29157357-8b8609ce-7de1-11e7-9bcf-bda863edfb92.png)

![2017-08-10 15 35 01](https://user-images.githubusercontent.com/2675542/29157355-8b59a62c-7de1-11e7-90f6-be58954ea3bb.png)

As you can see, mz_crc32 is missing in the lower one.

If you want example files, I can provide, but not today.

**Comment by lukaslueg** [comment]
I'm using flate2 to decompress lots and lots of data which has already gone by hash-checking, so the stream can be trusted to be free of corruption. Profiling shows `mz_crc32` to take upwards of 30% of cpu usage, which is quite a lot.

**Comment by alexcrichton** [comment]
I've implemented a SIMD version of CRC checksums in https://github.com/alexcrichton/flate2-rs/pull/172 which is over 50x faster than miniz, and in testing zlib is also over 2x faster than zlib as well. I'm gonna close this with that PR as hopefully it will obviate the need for a flag such as this.

**Comment by lukaslueg** [comment]
Since this saves a quite significant amount of CPU-time, we may have a bump in `flate2` and cargo's lockfile?

**Comment by alexcrichton** [comment]
Yep! Was just waiting on >24 hrs of fuzzing to finish to confirm nothing shows up

**Comment by kootenpv** [comment]
Another reason to ignore CRC32 would be to just read partial data. How to go about ignoring the CRC32 check?

**Comment by aconz2** [comment]
I came across this issue after seeing crc32 from CrcReader (`crc32fast::specialized::pclmulqdq::calculate` specifically) show up as 5% of my profile when using zlib-ng(-sys). After looking through the zlib-ng source, I was curious about their crc implementation (which conditionally runs (does not run with GzDecoder) if given the right inflateInit params and sees the gzip header) and came up with this trick to make zlib-ng track the crc. That value is accessible via the z_stream `adler` field. 

```rust
let mut decoder = bufread::ZlibDecoder::new_with_decompress(
    BufReader::new(File::open(file).unwrap()),
    Decompress::new_gzip(15),
);
```

(Note the file passed in is still a gzip file, we are only using ZlibDecoder because it takes a configurable Decompress)

Using this method to decompress, the runtime is almost identical with `crc32_fold_pclmulqdq_copy` from zlib-ng taking 5% instead of from crc32fast. Initially I thought it might be worth using their crc since they calculate while copying (though not all backends support this), but it seems not worthwhile.

To skip crc, I found you can do this:

```rust
let decoder = bufread::GzDecoder::new(BufReader::new(File::open(file).unwrap()));
assert!(decoder.header().is_some()); // none if EWOULDBLOCK (is there a way to spin until you read the header? maybe call fill before passing to GzDecoder)
let mut decoder = bufread::DeflateDecoder::new(decoder.into_inner());
```

so that GzDecoder reads the header and then have DeflateDecoder take over from there. DeflateDecoder will finish at the end of data and the inner reader will then read the remaining 8 bytes of crc32 and len32 (for single file gzip).
